### PR TITLE
Refactor gRPC path to use LoggingServiceV2Client.

### PR DESCRIPTION
**Pull request by qingling128**
This PR depends on https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/173.

And https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/174 depends on this PR.

### CODE REVIEW

**Comment by qingling128** [comment]
#173 has been merged. This PR has been rebased.

File: test/plugin/base_test.rb [hunk 1]
```diff
@@ -223,10 +223,10 @@ def test_one_log_with_invalid_json_credentials
     setup_gce_metadata_stubs
     ENV['GOOGLE_APPLICATION_CREDENTIALS'] = INVALID_CREDENTIALS[:path]
     setup_logging_stubs do
-      d = create_driver
-      d.emit('message' => log_entry(0))
-      exception_count = 0
       begin
+        exception_count = 0
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: Why did you need to move these in? Can `create_driver` or `emit` raise a `RuntimeError`? At least you should be able to move the initialization of `exception_count` back out...

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 2]
```diff
@@ -51,17 +52,15 @@ def test_server_error
...
+        d = create_driver(
+          USE_GRPC_CONFIG, 'test',
+          GRPCLoggingMockFailingService.new(code, message, @failed_attempts))
         # The API client should retry this once, then throw an exception which
         # gets propagated through the plugin
         d.emit('message' => log_entry(0))
         begin
           d.run
-        rescue GRPC::Cancelled
-          # No need to check the message -- we already know the code.
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Was this code wrong, or does it no longer apply?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 3]
```diff
@@ -212,33 +217,22 @@ def test_non_integer_timestamp
     use_grpc true
   ).freeze
 
-  @@mock_port = 0 # rubocop:disable Style/ClassVars
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Is this no longer needed because of the changes in this PR, or because the underlying issue has been fixed? If the latter, could we at least split it out into a separate commit?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 4]
```diff
@@ -249,21 +243,37 @@ def api_client
 
       # Here we have obtained the creds, but for the mock, we will leave the
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Does this comment still apply?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 5]
```diff
@@ -249,21 +243,37 @@ def api_client
...
+  class GRPCLoggingMockService <
+      Google::Cloud::Logging::V2::LoggingServiceV2Client
     def initialize(requests_received)
       super()
       @requests_received = requests_received
     end
 
-    def write_log_entries(request, _call)
+    # rubocop:disable Lint/UnusedMethodArgument
+    # rubocop:disable Metrics/ParameterLists
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: (a) You can exclude keyword arguments by setting `CountKeywordArgs` in `.rubocop.yml` to false.
(b) Is it time to re-generate `.rubocop_todo.yml`?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 6]
```diff
@@ -249,21 +243,37 @@ def api_client
...
+    def write_log_entries(entries,
+                          log_name: nil,
+                          resource: nil,
+                          labels: nil,
+                          partial_success: nil,
+                          options: nil)
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: If `options` is unused, do we even need to specify it?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 7]
```diff
@@ -249,21 +243,37 @@ def api_client
...
+      request = Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Should we not store `partial_success` in this request?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 8]
```diff
@@ -297,7 +307,12 @@ def initialize(code, message, failed_attempts)
 
     def write_log_entries(_request, _call)
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Is there a reason we're keeping this class as a subclass of `Google::Logging::V2::LoggingServiceV2::Service`, rather than making it a subclass of `Google::Cloud::Logging::V2::LoggingServiceV2Client`, same as `GRPCLoggingMockService`?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 9]
```diff
@@ -318,43 +333,22 @@ def _undefined
   end
 
   # Set up grpc stubs to mock the external calls.
-  # TODO(qingling128): Remove this comment after grpc/grpc#12506 is resolved.
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: The issue is still open. What changed on our side to no longer need this?

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 10]
```diff
@@ -318,43 +333,22 @@ def _undefined
...
-  # Due to a gRPC load balancing issue (grpc/grpc#12506), we have to use a
-  # different port each time we create a gRPC mock as a temporary workaround.
-  # Thus we can only create one driver in each setup_logging_stubs context.
-  def setup_logging_stubs(should_fail = false, code = 0, message = 'Ok')
-    # Save the mock host in an instance variable, so later on when creating the
-    # logging driver, we can refer to this host with exactly the same port.
-    @mock_host = generate_mock_host
-    srv = GRPC::RpcServer.new
+  def setup_logging_stubs
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Is this now effectively a no-op? Where are the actual stubs set up? Shouldn't that be here?

File: lib/fluent/plugin/out_google_cloud.rb [hunk 11]
```diff
@@ -595,44 +595,38 @@ def write(chunk)
...
-              increment_dropped_entries_count(entries_count)
-              @log.warn "Dropping #{entries_count} log message(s)",
-                        error: error.to_s, error_code: error.code.to_s
-            when GRPC::Core::StatusCodes::UNAUTHENTICATED
-              # Authorization error.
-              # These are usually solved via a `gcloud auth` call, or by
-              # modifying the permissions on the Google Cloud project.
+
+            # Most client errors indicate a problem with the request itself and
+            # should not be retried. Auth errors are usually solved via a
```
[the remark this answers is not in the document]
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: test/plugin/base_test.rb (hunk 1 above)
Comment (qingling128) [reply test/plugin/base_test.rb]: For grpc path, it used to raise a `RuntimeError('Unable to read the credential file')` at `d.run` with invalid json credentials. The new lib raises the same `RuntimeError` during `create_driver` instead.

File: test/plugin/test_out_google_cloud_grpc.rb (hunk 2 above)
Comment (qingling128) [reply test/plugin/test_out_google_cloud_grpc.rb]: Turned out that `GRPC::Cancelled` is a type of `GRPC::BadStatus`. So it was unnecessary to treat that one separately.

File: test/plugin/test_out_google_cloud_grpc.rb (hunk 3 above)
Comment (qingling128) [reply test/plugin/test_out_google_cloud_grpc.rb]: This is no longer needed because we are no longer mocking the RPC server directly. Instead we are mocking out the client library's `LoggingServiceV2Client`'s `write_log_entries` method. `LoggingServiceV2Client` wraps the RPC client stub.

File: test/plugin/test_out_google_cloud_grpc.rb (hunk 4 above)
Comment (qingling128) [reply test/plugin/test_out_google_cloud_grpc.rb]: Nope. Removed.

File: test/plugin/test_out_google_cloud_grpc.rb (hunk 6 above)
Comment (qingling128) [reply test/plugin/test_out_google_cloud_grpc.rb]: Yeah, the mock needs to match exactly the original method arguments.

**Comment by qingling128** [comment]
Gentle reminder: PTAL.

**Comment by qingling128** [comment]
Please ignore the commits from `Bump version to v0.6.9.` to `Audit, comment and fix gRPC status code handling.`

I'll rebase them once the PR is approved.

**Comment by qingling128** [comment]
Seems it's not very straight forward to figure out what actually is changed from github's view. I made a `lingshi-grpc-refactor-rebase` branch if that helps. The diff betwen `master` and that branch can be viewed at https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/192/files. Or maybe it's easier to follow up the code review there. Sorry for the inconvenience. Let me know if there is anything I could do to make it easier.

**Comment by qingling128** [comment]
Closing this PR because it's kind of messy merging in all the latest code in master. Follow up will be in https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/192.
